
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Installation &#8212; MINUUR - Microbial INsights Using Unmapped Reads</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Inputting Files To MINUUR" href="03_Inputting_Files.html" />
    <link rel="prev" title="Introduction" href="index.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
      
      
      <h1 class="site-logo" id="site-title">MINUUR - Microbial INsights Using Unmapped Reads</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="index.html">
                    Introduction
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Installation
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="03_Inputting_Files.html">
   Inputting Files To MINUUR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="04_Configuring_The_Workflow.html">
   Configuring the Pipeline
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="05_Running_MINUUR.html">
   Running MINUUR
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="06_Outputs_Explained.html">
   Outputs Explained
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="07_Weird_Things_Happening.html">
   Weird Things are Happening…What’s the Problem?
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/executablebooks/jupyter-book"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/executablebooks/jupyter-book/issues/new?title=Issue%20on%20page%20%2F02_Installation.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/02_Installation.md"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.md</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#conda">
   Conda
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snakemake">
   Snakemake
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#databases">
   Databases
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#installing-a-database">
   Installing a Database
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-host">
   A Host
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#getting-the-pipeline">
   Getting the Pipeline
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Installation</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#conda">
   Conda
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#snakemake">
   Snakemake
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#databases">
   Databases
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#installing-a-database">
   Installing a Database
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-host">
   A Host
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#getting-the-pipeline">
   Getting the Pipeline
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">#</a></h1>
<p>To use MINUUR, there are several things you will need to install: dependencies for the pipeline to run (conda and Snakemake) and several taxonomic databases for microbial classification.</p>
<section id="conda">
<h2>Conda<a class="headerlink" href="#conda" title="Permalink to this headline">#</a></h2>
<p>Conda is a package management system to install and manage contained software environemnts. MINUUR uses conda to create isolated software environemnts with their specific packages and versions to run, and is also required to install the next dependency - Snakemake. If conda isn’t installed on your system, please refer to the <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html">documentation</a> to  get started.</p>
</section>
<section id="snakemake">
<h2>Snakemake<a class="headerlink" href="#snakemake" title="Permalink to this headline">#</a></h2>
<p>Snakemake is a workflow manager to create reproducible analyses and was used to write MINUUR. Snakemake makes it easy to create complex and configurable workflows that should run seamlessly on other systems using isolated conda environments. Please follow the <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/getting_started/installation.html">documentation</a> on how to install Snakemake on your system.</p>
</section>
<section id="databases">
<h2>Databases<a class="headerlink" href="#databases" title="Permalink to this headline">#</a></h2>
<p>MINUUR uses large taxonomic databases to run several parts of its workflow. You can pick and choose which databases are relevent to your question at hand but the databases MINUUR requires to run a full analysis, or reproduce my analysis in this <a class="reference external" href="https://www.biorxiv.org/content/biorxiv/early/2022/08/11/2022.08.09.503283.full.pdf">paper</a> are:</p>
<ol class="simple">
<li><p>An indexed Kraken2 Database (alternatively a manually created Kraken2 database if you are interested in finding specific taxa). Used for classifying those unmapped reads.</p></li>
<li><p>An indexed Bracken Database. An extra step to reestimate relative abundance from your Kraken2 Classifications.</p></li>
<li><p>A MetaPhlAn3 Database - you can use this in parallel or alternatively to Kraken2 since MetaPhlAn3 uses marker genes instead of k-mers - an alternative way to classify your unmapped reads. I found Kraken2 to work better when classifying unmapped reads from mosquitoes.</p></li>
<li><p>A HUMAnN3 Database - Here a ChocoPhlAn database and UniRef90 database can be used to search for functional profiles from predicted taxa - this would be used in conjunction with MetaPhlAn3’s classifications.</p></li>
</ol>
<p>That’s quite a few databases…</p>
<p>You can pick and choose what analysis to do and is relevent to you, so feel free to omit any database you are not interested in. Certain steps can be ignored using the configuration file (more on this later, but to skip ahead look <a class="reference internal" href="04_Configuring_The_Workflow.html"><span class="doc std std-doc">here</span></a>).</p>
<p>While obtaining all of this from external sources is a bit of a faff, the brilliant people hosting these databases will update and maintain these databases.</p>
</section>
<section id="installing-a-database">
<h2>Installing a Database<a class="headerlink" href="#installing-a-database" title="Permalink to this headline">#</a></h2>
<p>If the sever you are using already has one of these databases installed and ready to use, do nothing. You will specify the paths to these database in the later steps when configuring your workflow.</p>
<p>If you require these databases, please go to the relevent pages and install them from there - links to each are:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://benlangmead.github.io/aws-indexes/k2">Kraken2</a></p></li>
<li><p><a class="reference external" href="https://benlangmead.github.io/aws-indexes/k2">Bracken</a> (the Kraken and Bracken index can be used for the same analysis if downloaded from this repository)</p></li>
<li><p><a class="reference external" href="https://github.com/biobakery/MetaPhlAn/wiki/MetaPhlAn-3.0">MetaPhlAn3</a></p></li>
<li><p><a class="reference external" href="https://github.com/biobakery/humann">Humann3</a></p></li>
</ul>
</section>
<section id="a-host">
<h2>A Host<a class="headerlink" href="#a-host" title="Permalink to this headline">#</a></h2>
<p>MINUUR tries to recover metagenomic information from a host in question. In my case, this was from a load of mozzies. In your case…it might be something cooler. To account for different hosts, MINUUR is flexible when inputting a target host of interest. There’s a caveat, however. Your host of interest must be indexed using bowtie2. Follow the instructions <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#the-bowtie2-build-indexer">here</a>. If you make your way to the top of the link provided, you will find some ready-made indexes of some common organisms (maybe yours is there!). If not, you’ll need to index a chromosome-assembled fasta file of your organism in question and specify a path to your indexed reference in the configuration file.</p>
</section>
<section id="getting-the-pipeline">
<h2>Getting the Pipeline<a class="headerlink" href="#getting-the-pipeline" title="Permalink to this headline">#</a></h2>
<p>When you’re ready, you can acquire the pipeline using <code class="docutils literal notranslate"><span class="pre">git</span> <span class="pre">clone</span> <span class="pre">https://github.com/aidanfoo96/MINUUR/</span></code> into your chosen working directory. Once you’ve cloned the pipeline, navigate to <code class="docutils literal notranslate"><span class="pre">cd</span> <span class="pre">MINUUR/workflow</span></code>. This is the reference point where you will run the pipeline.</p>
<p>Try typing <code class="docutils literal notranslate"><span class="pre">snakemake</span> <span class="pre">-np</span></code> into your terminal once you’ve navigated to <code class="docutils literal notranslate"><span class="pre">MINUUR/workflow</span></code> … you’ll praobbly get an error. That’s because we need to specify some data to analyse! On to the next section!</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="index.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Introduction</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="03_Inputting_Files.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Inputting Files To MINUUR</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Aidan Foo<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>